File: Source/Entities/DashDirController.cs
using System;
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;
using static Celeste.Mod.WarlockHelper.Utils.Dir8;

namespace Celeste.Mod.WarlockHelper.Entities;

[CustomEntity("WarlockHelper/DashDirController")]
[Tracked]

public class DashDirController : Entity
{
    private readonly Func<Player,Vector2> DashDirFunc;

    private Player Player;
    public DashDirController(EntityData data, Vector2 _)
    {
        bool overrideNeutral = data.Bool("overrideNeutral");
        Vector2 neutralLeft = data.Vector2Grouped("neutralLeft"), neutralRight=data.Vector2Grouped("neutralRight");
        if (data.Int("_mode") == 0)
        {
            Utils.Matrix2x2 matrix = new (data.FloatArray("direction",6,6,[1,0,0,1,0,0]));
            if (!overrideNeutral)
            {
                neutralLeft = LEFT.Transform(matrix);
                neutralRight = RIGHT.Transform(matrix);
            }
            DashDirFunc = (player =>
            {
                if (Input.Aim.Value == Vector2.Zero)
                {
                    return player.Facing == Facings.Left ? neutralLeft:neutralRight;
                }
                return Input.GetAimVector(player.Facing).Transform(matrix);
            });
            
        }
        else
        {
            Vector2[] dirs =
            [
                data.Vector2Grouped("right",RIGHT),
                data.Vector2Grouped("downRight",DOWNRIGHT),
                data.Vector2Grouped("down",DOWN),
                data.Vector2Grouped("downLeft",DOWNLEFT),
                data.Vector2Grouped("left",LEFT),
                data.Vector2Grouped("upLeft",UPLEFT),
                data.Vector2Grouped("up",UP),
                data.Vector2Grouped("upRight",UPRIGHT)
            ];
            if (!overrideNeutral)
            {
                neutralLeft = dirs[LEFT.Angle8()];
                neutralRight = dirs[RIGHT.Angle8()];
            }
            DashDirFunc = (player =>
            {
                if (Input.Aim.Value == Vector2.Zero)
                {
                    return player.Facing == Facings.Left ? neutralLeft:neutralRight;
                }
                int angle = Input.GetAimVector(player.Facing).Angle8();
                return dirs[angle];
            });
        }
    }

    public DashDirController(Func<Player, Vector2> dashDirFunc)
    {
        DashDirFunc = dashDirFunc;
    }

    public override void Awake(Scene scene)
    {
        base.Awake(scene);
        Player = scene.Tracker.GetEntity<Player>();
        Player.SetDefaultDashDir(DashDirFunc);
        Debug.Log(
            $"Dash Direction Controller from room {SourceData.Level.Name} activated",LogLevel.Verbose,"DashDirController.Removed");
    }

    public override void Removed(Scene scene)
    {
        base.Removed(scene);
        foreach (Entity entity in scene.Tracker.GetEntities<DashDirController>())
        {
            if (entity != this)
            {
                return; //reset dashDirection only if there aren't any other controllers
            }
        }
        Player.SetDefaultDashDir(null);
        Debug.Log(
            $"Dash Direction Controller from room {SourceData.Level.Name} deactivated", LogLevel.Verbose,
            "DashDirController.Removed");
    }
}

File: Loenn/entities/dashDirController.lua
local mods = require("mods")
local wsUtils = mods.requireFromPlugin("libraries.utils")
local ftype = mods.requireFromPlugin("libraries.fieldTypes")

--local logging = require("logging")

local dashDirController = {}
dashDirController.name = "WarlockHelper/DashDirController"
dashDirController.depth = -1000000
dashDirController.texture = "objects/WarlockHelper/dashDirController"

function dashDirController.ignoredFields (entity)
	local allFields = {
	"_name", "_id", "originX", "originY",
	"_mode",
	"right",
	"downRight",
	"down",
	"downLeft",
	"left",
	"upLeft",
	"up",
	"upRight",
	"neutralLeft",
	"neutralRight",
	"direction",
	"overrideNeutral",
	}
	local visible = {}
	local function show(...)
		local arg = {...}
		for _,v in ipairs(arg) do
			table.insert(visible,v)
		end
	end
	show("overrideNeutral")
	if entity.overrideNeutral then
		show("neutralLeft","neutralRight")
	end
	if entity._mode == 0 then
		show("direction")
	end
	if entity._mode == 1 then
		show(
			"right",
			"downRight",
			"down",
			"downLeft",
			"left",
			"upLeft",
			"up",
			"upRight"
		)
	end
	return wsUtils.setDiff(allFields,visible)
end

dashDirController.fieldOrder = {
	"x","y",
	"right",
	"downRight",
	"down",
	"downLeft",
	"left",
	"upLeft",
	"up",
	"upRight",
	"neutralLeft",
	"neutralRight",
	"direction",
	"overrideNeutral",
}
dashDirController.fieldInformation = {
	direction = ftype.matrix,
	right=ftype.vector2,
	downRight=ftype.vector2,
	down=ftype.vector2,
	downLeft=ftype.vector2,
	left=ftype.vector2,
	upLeft=ftype.vector2,
	up=ftype.vector2,
	upRight=ftype.vector2,
	neutralLeft = ftype.vector2,
	neutralRight = ftype.vector2,
}
	dashDirController.placements = {
		{
			name = "normal",
			data = {
				_mode = 0,
				direction = ftype.matrix.default,
				neutralLeft = "-1,0",
				neutralRight = "1,0",
				overrideNeutral=false,
			}
		},
		{
			name = "map",
			data = {
				_mode = 1,
				right="1,0",
				downRight="0.70710677,0.70710677",
				down="0,1",
				downLeft="-0.70710677,0.70710677",
				left="-1,0",
				upLeft="-0.70710677,-0.70710677",
				up="0,-1",
				upRight="0.70710677,-0.70710677",
				neutralLeft = "-1,0",
				neutralRight = "1,0",
				overrideNeutral = false,
			}
		}
	}

	return dashDirController